### ISSUE TITLE
Announcement: Future development of lib3mf

### ISSUE THREAD

**Issue by martinweismann** [3MFConsortium/lib3mf#127]
Dear contributors and users of lib3mf,

we just released a preview for **major version 2** of lib3mf: 
https://github.com/3MFConsortium/lib3mf/releases/tag/v2.0.0-alpha
to gather feedback and give you the opportunity report bugs early on.

Apart from the improvements visible to _users_ of the library, like native C++-,Python- and Go-language bindings or multiproperty-support and a more contemporary documentation (https://lib3mf.readthedocs.io/en/latest/), the way the library is built has changed, too:

lib3mf now makes use of the https://github.com/Autodesk/AutomaticComponentToolkit/, an open-source code generator for shared library components that helps us save _a lot_ of work when creating the interface, the language bindings and the documentation for lib3mf.

The code of lib3mf v2 currently lives on https://github.com/3MFConsortium/lib3mf/tree/lib3mf-v2-develop, before it will eventually be merged into the develop-branch and finally released as official version 2.0.0. In general, future development of features should be based on version 2 of lib3mf. On the same note, developers integrating lib3mf into their code are advised to switch to v2 at some point, too.

I hope we did not break to much in the transition from v1.* to v2 but I am confident the work will pay off in the end.

Best regards
Martin Weismann

**Comment by shahchintan14** [comment]
Is there an API to set default property of a mesh object? Couldn't find one.

**Comment by artemzakharov** [comment]
Are there any plans to add converters other than STL? Converting to and from common formats is currently by far the biggest pain point of working with the format, and while the STL converter is great at what it does, it kind of defeats the purpose of using 3MF to begin with - having STL as the intermediary between 3MF and all other file formats means things like textures and materials are invariably lost, as STLs can't hold that kind of information. A converter for going directly from OBJ to 3MF while preserving all of the extra information would be very useful!

**Comment by martinweismann** [comment]
@shahchintan14 : that's right, lib3mf is currently missing the Set/GetDefaultProperty-functionality. We shall add that before properly releasing v2.

**Comment by martinweismann** [comment]
@artemzakharov , that's an interesting point.

lib3mf is not primarily meant to be a converter. No other file format that is used in additive manufacturing can hold all of the information that can be contained in a 3MF-file (slices, beamlattices, assemblies, ...) and I assume we are on the same page that STL should be avoided as much as possible): Writing a filetype other than 3MF from lib3mf will always loose some information.

Instead, client applications (software, services, printers, ...) should use 3MF as interchange format. Hence, the main goal of lib3mf is to make it easy for these client applications to obtain all information from the 3MF file into their data model or to export all their relevant additional information as 3MF via lib3mf's API.

Having said that, additional writers and especially _readers_ (e.g. for obj, which has at least _some_ of 3MF's features) would be a great utility functionality for lib3mf and we might add more readers at some point. If you ever work on a decent obj-reader, it would be great if you could contribute that to lib3mf!

**Comment by artemzakharov** [comment]
@martinweismann In my particular use case, I'm actually distributing the models to hobbyists for personal printing, so there's no proprietary format for me to worry about - I just distribute the 3MF itself since it's a great format and is better suited for printing than OBJ, which my files originally come as in 99% of cases. Unfortunately, I haven't been able to find _any_ other converters whatsoever that import or export 3MF, so using STL as the intermediary with lib3mf is my only option right now, as I lack the resources to personally write an OBJ reader & exporter at the moment.

Keep us posted in case you do decide to add them, though!

**Comment by martinweismann** [comment]
Dear all,

we just released a second preview for major version 2 of lib3mf:
https://github.com/3MFConsortium/lib3mf/releases/tag/v2.0.0-beta

It contains fixes and improvements over v2.0.0-alpha based on your feedback over the last months. Thank you for your great feedback!

Again, if you find bugs in v2.0.0-beta or have urgent feature requests, please reach out as usual.

We hope to release v2.0.0 for production in ~2 months.

Cheers
Martin

**Comment by shahchintan14** [comment]
Hi @martinweismann ,
Thank you for quick response to the feedback.
Regarding the production version of 2.0, is there any chance to release it by July end? (In case there are not many issues and general response is good?)

**Comment by martinweismann** [comment]
It certainly depends on the feedback. It is possible that it will work out that way.

Considering this is a major update, it should be as clean as possible so that we do not start off with a new list of  required changes that would break backward-compatibility.

**Comment by martinweismann** [comment]
Dear contributors and users of lib3mf,

after two pre-releases, we just released **major version 2** of lib3mf for use in production!
https://github.com/3MFConsortium/lib3mf/releases/tag/v2.0.0

**Thank you** to everyone who contributed to this gigantic release over the last 9 months!

Future developments and PRs should target the [develop](https://github.com/3MFConsortium/lib3mf/tree/develop)-branch.  
There is a chance that we will publish another release on the version 1.*-line of lib3mf. The development of this potential release is tracked on the [develop_v1_legacy](https://github.com/3MFConsortium/lib3mf/tree/develop_v1_legacy)-branch.
